Set priority for eth transactions

yihuang opened the pull request:
## Description

Set the tx priority to the lowest priority in the messages.

<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->


<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

yihuang commented on CHANGELOG.md:
```diff
@@ -52,6 +52,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 * (feemarket) [\#1165](https://github.com/evmos/ethermint/pull/1165) Add hint in specs about different gas terminology for gas in Cosmos and Ethereum.
 * (rpc) [\#1169](https://github.com/evmos/ethermint/pull/1169) Remove unnecessary queries from `getBlockNumber` function
 * (ante) [1208](https://github.com/evmos/ethermint/pull/1208) Change default `maxGasWanted` value 
+* (ante) []() Set mempool priority to evm transactions.
```
```suggestion
* (ante) [#1214](https://github.com/evmos/ethermint/pull/1214) Set mempool priority to evm transactions.
```

facs95 commented:
Unit tests are failing @yihuang 🙏

fedekunze commented on x/evm/keeper/utils.go:
```diff
@@ -48,41 +55,53 @@ func (k Keeper) DeductTxCostsFromUserBalance(
+	if baseFee != nil {
+		tipPrice = new(big.Int).Sub(tipPrice, baseFee)
+	}
+	priorityBig := new(big.Int).Quo(tipPrice, DefaultPriorityReduction.BigInt())
```
would be good to add tests with different priorities

fedekunze reviewed (commented):
LGTM, would be great to add more integration tests and update the specification. Some remarks

1. I see the case for some applications that might not want to enable the prioritized mempool so in that scenario, we should add a parameter to the EVM to disable tx prioritization
2. Do we need to update the Cosmos ante handler to enable prioritization?

yihuang commented:
> * I see the case for some applications that might not want to enable the prioritized mempool so in that scenario, we should add a parameter to the EVM to disable tx prioritization

The validator is still free to choose the `v0` mempool which will ignore the priority field.

> * Do we need to update the Cosmos ante handler to enable prioritization?

The cosmos ante handler has a default mechanism to extract prioritization from fees directly, in another PR we'll customize that one with feemarket.

fedekunze commented on x/evm/keeper/utils.go:
```diff
@@ -14,20 +15,26 @@ import (
+// DefaultPriorityReduction is the default amount of price values required for 1 unit of priority.
+// Because priority is `int64` while price is `big.Int`, it's necessary to scale down the range to keep it more pratical.
+// The default value is the same as the `sdk.DefaultPowerReduction`.
+var DefaultPriorityReduction = sdk.DefaultPowerReduction
```
still not sure why this needs to be a separate variable

yihuang replied:
Do you mean to use `sdk.DefaultPowerReduction` directly? I think they serve a different purpose, the user may want to scale to different degree?

fedekunze reviewed (changes requested):
We should add specific integration tests for tx prioritization before we can merge this:

- legacy and access list txs
- dynamic fee tx with 0 and  > 0 priority tip

We also need to update the specification for the EVM with more details about tx prioritization for Ethereum msgs

yihuang commented:
> We should add specific integration tests for tx prioritization before we can merge this:
> 
> * legacy and access list txs
> * dynamic fee tx with 0 and  > 0 priority tip

added a priority integration test, which checked that a higher priority tx gets included first even if sent later.

> 
> We also need to update the specification for the EVM with more details about tx prioritization for Ethereum msgs

added sth in feemarket's spec in the ante handlers section.

yihuang commented on x/evm/keeper/utils.go:
```diff
@@ -14,20 +15,26 @@ import (
 	denom string,
 	homestead, istanbul, london bool,
-) (sdk.Coins, error) {
+) (sdk.Coins, int64, error) {
```
done

fedekunze commented on go.mod:
```diff
@@ -191,4 +191,7 @@ replace (
 	github.com/gin-gonic/gin => github.com/gin-gonic/gin v1.7.0
 	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
+
+	// TODO: remove after fixed https://github.com/cosmos/cosmos-sdk/issues/11364
+	github.com/zondax/hid => github.com/zondax/hid v0.9.0
```
why do we need this here?

fedekunze commented on tests/integration_tests/test_priority.py:
```diff
@@ -0,0 +1,112 @@
+    assert all(receipt.status == 1 for receipt in receipts), "expect all txs success"
+
+    # the later txs should be included earlier because of higher priority
+    # FIXME there's some non-deterministics due to mempool logic
```
this should be addressed before merging this

yihuang commented on tests/integration_tests/test_priority.py:
```diff
@@ -0,0 +1,114 @@
+    txhashes = [w3.eth.send_raw_transaction(tx.rawTransaction) for tx in signed]
+
+    receipts = [w3.eth.wait_for_transaction_receipt(txhash) for txhash in txhashes]
+    print(receipts)
```
```suggestion
```

fedekunze reviewed (approved):
ACK. @facs95 @alexanderbez can you review?
